File: Dsp-lab/Experiment-5/codes/main.m
function main()

N = 39;

%lpf
wc = pi/2;
fc= 400;
fs_lpf = (2*pi*fc)/(wc);
h1 = lpf_func(fc,fs_lpf,N).*hw(N);
fvtool(h1);
disp('h1');
disp(h1);
x1 = 1 :N;
figure
stem(x1,h1)
xlabel('N-sample')
ylabel('lpf x wd (Impulse)')
grid on;

%bpf
fc1 = 500;
fc2 = 1200;
fs_bpf = 6000;
h2 = bpf_func(fc1,fc2,fs_bpf,N).*hw(N);
fvtool(h2);
disp('h2');
disp(h2);
x2 = 1:N;
figure
stem(x2,h2)
xlabel('N-sample')
ylabel('bpf x wd (Imppulse)')
grid on;

end

function y = lpf_func(fc,fs,N)

wc = (2*pi*fc)/(fs);
n = 1;
y = zeros(1,N-1);
for k = -(N-1)/2:(N-1)/2 
    if k == 0
        y(n) = wc/pi;
    else
        y(n) = sin((wc*k))/(pi*k);
    end
    n = n + 1;
end
end

function y = bpf_func(fc1,fc2,fs,N) %bandpass

wc1 = (2*pi*fc1)/(fs);
wc2 = (2*pi*fc2)/(fs);
n = 1;
y = zeros(1,N-1);
for k = -(N-1)/2: (N-1)/2
    if k == 0
        y(n) = (wc2 - wc1)/pi;
    else
        y(n) = (sin((wc2*k))/(pi*k)) - (sin((wc1*k))/(pi*k));
    end
    n = n + 1;
end
end

function y = hw(N) %window function

n = 1;
y = zeros(1,N);
for k = 0  : (N-1)
    if k >= 0 && k <= (N-1)
        y(n) = 0.54 - 0.46*cos((2*pi*k)/(N-1));
    else 
        y(n) = 0;
    end
    n = n + 1;
end
end

File: Dsp-lab/Experiment-5/codes/main.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define N 39
#define M_PI 3.14159265358979323846

void lpf_func(double fc, double fs, double* h) {
    double wc = (2 * M_PI * fc) / fs;
    for (int k = -(N - 1) / 2, n = 0; k <= (N - 1) / 2; ++k, ++n) {
        if (k == 0) {
            h[n] = (wc / M_PI) ;
        } else {
            h[n] = (sin(wc * k) / (M_PI * k)) ;
        }
    }
}

void bpf_func(double fc1, double fc2, double fs, double* h) {
    double wc1 = (2 * M_PI * fc1) / fs;
    double wc2 = (2 * M_PI * fc2) / fs;
    for (int k = -(N - 1) / 2, n = 0; k <= (N - 1) / 2; ++k, ++n) {
        if (k == 0) {
            h[n] = ((wc2 - wc1) / M_PI) ;
        } else {
            h[n] = ((sin(wc2 * k) / (M_PI * k)) - (sin(wc1 * k) / (M_PI * k)));
        }
    }
}

double hw(int i) {
    double window[N];
    for (int k = 0; k < N; ++k) {
        window[k] = 0.54 - 0.46 * cos((2 * M_PI * k) / (N - 1));
    }
    return window[i];
}

int main() {
    double h1[N], h2[N];
    double fc = 400, fs_lpf = 1600;
    lpf_func(fc, fs_lpf, h1);

    double fc1 = 500, fc2 = 1200, fs_bpf = 6000;
    bpf_func(fc1, fc2, fs_bpf, h2);

    
    printf("h1:\n");
    for (int i = 0; i < N; ++i) {
        printf("%f ", h1[i]*hw(i));
    }
    printf("\n");

    printf("h2:\n");
    for (int i = 0; i < N; ++i) {
        printf("%f ", h2[i]*hw(i));
    }
    printf("\n");

    return 0;
}
